File: app/src/main/java/com/micewine/emu/fragments/VirtualControllerPresetManagerFragment.java
package com.micewine.emu.fragments;

import static com.micewine.emu.activities.MainActivity.getNativeResolution;
import static com.micewine.emu.activities.MainActivity.gson;
import static com.micewine.emu.activities.MainActivity.preferences;
import static com.micewine.emu.activities.PresetManagerActivity.SELECTED_VIRTUAL_CONTROLLER_PRESET;
import static com.micewine.emu.adapters.AdapterPreset.selectedPresetId;
import static com.micewine.emu.fragments.CreatePresetFragment.VIRTUAL_CONTROLLER_PRESET;
import static com.micewine.emu.views.VirtualKeyboardInputCreatorView.GRID_SIZE;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.reflect.TypeToken;
import com.micewine.emu.R;
import com.micewine.emu.adapters.AdapterPreset;
import com.micewine.emu.views.VirtualKeyboardInputView;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

public class VirtualControllerPresetManagerFragment extends Fragment {
    public static boolean editShortcut;

    public VirtualControllerPresetManagerFragment(boolean editShortcut) {
        initialize(editShortcut);
    }

    private static RecyclerView recyclerView;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_general_settings, container, false);
        recyclerView = rootView.findViewById(R.id.recyclerViewGeneralSettings);

        initialize(editShortcut);
        setAdapter();

        return rootView;
    }

    private void setAdapter() {
        recyclerView.setAdapter(new AdapterPreset(presetsAdapterList, requireContext(), requireActivity().getSupportFragmentManager()));

        presetsAdapterList.clear();
        presetsList.forEach((i) -> addToAdapter(i.name));
    }

    private void addToAdapter(String name) {
        presetsAdapterList.add(
                new AdapterPreset.Item(name, VIRTUAL_CONTROLLER_PRESET, true, false)
        );
    }

    private static ArrayList<VirtualControllerPreset> presetsList;
    private static final ArrayList<AdapterPreset.Item> presetsAdapterList = new ArrayList<>();
    private final static Type listType = new TypeToken<VirtualControllerPreset>() {}.getType();

    public static void initialize(boolean editable) {
        presetsList = getVirtualControllerPresets();
        editShortcut = editable;
    }

    public static VirtualControllerPreset getVirtualControllerPreset(String name) {
        int index = IntStream.range(0, presetsList.size()).filter(i -> presetsList.get(i).name.equals(name)).findFirst().orElse(-1);
        if (index == -1) return null;

        return presetsList.get(index);
    }

    public static void putVirtualControllerPreset(String name, String resolution,
                                                  ArrayList<VirtualKeyboardInputView.VirtualButton> buttonList,
                                                  ArrayList<VirtualKeyboardInputView.VirtualAnalog> analogList,
                                                  ArrayList<VirtualKeyboardInputView.VirtualDPad> dpadList
    ) {
        int index = IntStream.range(0, presetsList.size()).filter(i -> presetsList.get(i).name.equals(name)).findFirst().orElse(-1);
        if (index == -1) return;

        presetsList.set(index, new VirtualControllerPreset(name, resolution));

        buttonList.forEach((i) -> presetsList.get(index).buttons.add(i));
        analogList.forEach((i) -> presetsList.get(index).analogs.add(i));
        dpadList.forEach((i) -> presetsList.get(index).dpads.add(i));

        saveVirtualControllerPresets();
    }

    public static void addVirtualControllerPreset(Context context, String name) {
        boolean alreadyExists = presetsList.stream().anyMatch(i -> i.name.equals(name));
        if (alreadyExists) {
            Toast.makeText(context, context.getString(R.string.executable_already_added), Toast.LENGTH_SHORT).show();
            return;
        }

        presetsList.add(
                new VirtualControllerPreset(name)
        );
        presetsAdapterList.add(
                new AdapterPreset.Item(name, VIRTUAL_CONTROLLER_PRESET, true, editShortcut)
        );

        recyclerView.post(() -> {
            RecyclerView.Adapter<?> adapter = recyclerView.getAdapter();
            if (adapter != null) {
                adapter.notifyItemInserted(presetsList.size());
            }
        });

        saveVirtualControllerPresets();
    }

    public static void deleteVirtualControllerPreset(String name) {
        int index = IntStream.range(0, presetsList.size()).filter(i -> presetsList.get(i).name.equals(name)).findFirst().orElse(-1);

        presetsList.remove(index);
        presetsAdapterList.remove(index);

        if (index == selectedPresetId) {
            SharedPreferences.Editor editor = preferences.edit();

            editor.putString(SELECTED_VIRTUAL_CONTROLLER_PRESET, presetsList.get(0).name);
            editor.apply();
        }

        recyclerView.post(() -> {
            RecyclerView.Adapter<?> adapter = recyclerView.getAdapter();
            if (adapter != null) {
                adapter.notifyItemRemoved(index);

                if (index == selectedPresetId) {
                    adapter.notifyItemChanged(0);
                }
            }
        });

        saveVirtualControllerPresets();
    }

    public static void renameVirtualControllerPreset(String name, String newName) {
        int index = IntStream.range(0, presetsList.size()).filter(i -> presetsList.get(i).name.equals(name)).findFirst().orElse(-1);

        presetsList.get(index).name = newName;
        presetsAdapterList.get(index).setTitleSettings(newName);

        recyclerView.post(() -> {
            RecyclerView.Adapter<?> adapter = recyclerView.getAdapter();
            if (adapter != null) {
                adapter.notifyItemChanged(index);
            }
        });

        saveVirtualControllerPresets();
    }

    public static boolean importVirtualControllerPreset(Context context, File file) {
        List<String> lines;
        try {
            lines = Files.readAllLines(file.toPath());
        } catch (IOException e) {
            return false;
        }

        if (lines.size() < 2) return false;

        String type = lines.get(0);

        if (!type.equals("virtualControllerPreset")) return false;

        String json = lines.get(1);

        VirtualControllerPreset preset = gson.fromJson(json, listType);
        boolean autoAdjust = json.contains("resolution");

        String presetName = preset.name;
        int count = 1;

        while (true) {
            String currentName = presetName;
            if (presetsList.stream().anyMatch(p -> p.name.equals(currentName))) {
                presetName = preset.name + "-" + count++;
                continue;
            }
            break;
        }

        preset.name = presetName;

        if (autoAdjust && !preset.resolution.isEmpty()) {
            String nativeResolution = getNativeResolution(context);

            if (!nativeResolution.equals(preset.resolution)) {
                String[] nativeResolutionSplit = nativeResolution.split("x");
                String[] presetResolutionSplit = preset.resolution.split("x");

                float nativeResolutionWidth = Float.parseFloat(nativeResolutionSplit[0]);
                float nativeResolutionHeight = Float.parseFloat(nativeResolutionSplit[1]);

                float presetResolutionWidth = Float.parseFloat(presetResolutionSplit[0]);
                float presetResolutionHeight = Float.parseFloat(presetResolutionSplit[1]);

                float multiplierWidth = (nativeResolutionWidth / presetResolutionWidth) * 100F;
                float multiplierHeight = (nativeResolutionHeight / presetResolutionHeight) * 100F;

                preset.buttons.forEach((i) -> {
                    i.x = Math.round(i.x / 100F * multiplierWidth / GRID_SIZE) * (float) GRID_SIZE;
                    i.y = Math.round(i.y / 100F * multiplierHeight / GRID_SIZE) * (float) GRID_SIZE;
                });
                preset.analogs.forEach((i) -> {
                    i.x = Math.round(i.x / 100F * multiplierWidth / GRID_SIZE) * (float) GRID_SIZE;
                    i.y = Math.round(i.y / 100F * multiplierHeight / GRID_SIZE) * (float) GRID_SIZE;
                });
                preset.dpads.forEach((i) -> {
                    i.x = Math.round(i.x / 100F * multiplierWidth / GRID_SIZE) * (float) GRID_SIZE;
                    i.y = Math.round(i.y / 100F * multiplierHeight / GRID_SIZE) * (float) GRID_SIZE;
                });
            }
        }

        presetsList.add(preset);
        presetsAdapterList.add(
                new AdapterPreset.Item(preset.name, VIRTUAL_CONTROLLER_PRESET, true, false)
        );

        if (recyclerView != null) {
            recyclerView.post(() -> {
                RecyclerView.Adapter<?> adapter = recyclerView.getAdapter();
                if (adapter != null) {
                    adapter.notifyItemInserted(presetsList.size());
                }
            });
        }

        saveVirtualControllerPresets();

        return true;
    }

    public static void exportVirtualControllerPreset(String name, File file) {
        int index = IntStream.range(0, presetsList.size()).filter(i -> presetsList.get(i).name.equals(name)).findFirst().orElse(-1);

        try (FileWriter writer = new FileWriter(file)) {
            writer.write("virtualControllerPreset\n" + gson.toJson(presetsList.get(index)));
        } catch (IOException ignored) {
        }
    }

    private static void saveVirtualControllerPresets() {
        SharedPreferences.Editor editor = preferences.edit();

        editor.putString("virtualControllerPresetList", gson.toJson(presetsList));
        editor.apply();
    }

    public static ArrayList<VirtualControllerPreset> getVirtualControllerPresets() {
        String json = preferences.getString("virtualControllerPresetList", "");
        Type listType = new TypeToken<ArrayList<VirtualControllerPreset>>() {}.getType();
        ArrayList<VirtualControllerPreset> loadedList = gson.fromJson(json, listType);

        return (loadedList != null ? loadedList : new ArrayList<>());
    }

    public static class VirtualControllerPreset {
        public String name;
        public String resolution;
        public ArrayList<VirtualKeyboardInputView.VirtualAnalog> analogs;
        public ArrayList<VirtualKeyboardInputView.VirtualButton> buttons;
        public ArrayList<VirtualKeyboardInputView.VirtualDPad> dpads;

        public VirtualControllerPreset(String name) {
            this.name = name;
            this.buttons = new ArrayList<>();
            this.analogs = new ArrayList<>();
            this.dpads = new ArrayList<>();
        }

        public VirtualControllerPreset(String name, String resolution) {
            this.name = name;
            this.resolution = resolution;
            this.buttons = new ArrayList<>();
            this.analogs = new ArrayList<>();
            this.dpads = new ArrayList<>();
        }

        public String getName() {
            return name;
        }
    }
}

File: app/src/main/java/com/micewine/emu/fragments/VirtualControllerPresetManagerFragment.kt
package com.micewine.emu.fragments

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.edit
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.reflect.TypeToken
import com.micewine.emu.R
import com.micewine.emu.activities.MainActivity.Companion.getNativeResolution
import com.micewine.emu.activities.MainActivity.Companion.gson
import com.micewine.emu.activities.MainActivity.Companion.preferences
import com.micewine.emu.activities.PresetManagerActivity.Companion.SELECTED_VIRTUAL_CONTROLLER_PRESET
import com.micewine.emu.adapters.AdapterPreset
import com.micewine.emu.adapters.AdapterPreset.Companion.selectedPresetId
import com.micewine.emu.fragments.CreatePresetFragment.Companion.VIRTUAL_CONTROLLER_PRESET
import com.micewine.emu.views.VirtualKeyboardInputCreatorView.Companion.GRID_SIZE
import com.micewine.emu.views.VirtualKeyboardInputView
import java.io.File
import kotlin.math.roundToInt

class VirtualControllerPresetManagerFragment(private val editShortcut: Boolean) : Fragment() {
    private var rootView: View? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        rootView = inflater.inflate(R.layout.fragment_general_settings, container, false)
        recyclerView = rootView?.findViewById(R.id.recyclerViewGeneralSettings)

        initialize(editShortcut)
        setAdapter()

        return rootView
    }

    private fun setAdapter() {
        recyclerView?.setAdapter(
            AdapterPreset(presetListAdapters, requireContext(), requireActivity().supportFragmentManager)
        )

        presetListAdapters.clear()
        presetList.forEach {
            addToAdapter(it.name, VIRTUAL_CONTROLLER_PRESET, true)
        }
    }

    private fun addToAdapter(titleSettings: String, type: Int, userPreset: Boolean) {
        presetListAdapters.add(
            AdapterPreset.Item(titleSettings, type, userPreset, editShortcut)
        )
    }

    companion object {
        private var recyclerView: RecyclerView? = null
        private val presetListAdapters: MutableList<AdapterPreset.Item> = mutableListOf()
        private var presetList: MutableList<VirtualControllerPreset> = mutableListOf()
        private var editShortcut: Boolean = false
        private val listType = object : TypeToken<VirtualControllerPreset>() {}.type

        fun initialize(editable: Boolean = false) {
            presetList = getVirtualControllerPresets()
            editShortcut = editable
        }

        fun getVirtualControllerPreset(name: String): VirtualControllerPreset? {
            val index = presetList.indexOfFirst { it.name == name }
            if (index == -1) return null

            return presetList[index]
        }

        fun putVirtualControllerPreset(name: String, resolution: String, buttonList: MutableList<VirtualKeyboardInputView.VirtualButton>, analogList: MutableList<VirtualKeyboardInputView.VirtualAnalog>, dpadList: MutableList<VirtualKeyboardInputView.VirtualDPad>) {
            val index = presetList.indexOfFirst { it.name == name }
            if (index == -1) return

            presetList[index] = VirtualControllerPreset(name, resolution, mutableListOf(), mutableListOf(), mutableListOf())

            buttonList.forEach {
                presetList[index].buttons.add(it)
            }
            analogList.forEach {
                presetList[index].analogs.add(it)
            }
            dpadList.forEach {
                presetList[index].dpads.add(it)
            }

            saveVirtualControllerPresets()
        }

        fun addVirtualControllerPreset(context: Context, name: String) {
            if (presetListAdapters.firstOrNull { it.titleSettings == name } != null) {
                Toast.makeText(context, context.getString(R.string.executable_already_added), Toast.LENGTH_LONG).show()
                return
            }

            val defaultPreset = VirtualControllerPreset(name, "", mutableListOf(), mutableListOf(), mutableListOf())

            presetList.add(defaultPreset)
            presetListAdapters.add(
                AdapterPreset.Item(name, VIRTUAL_CONTROLLER_PRESET, true, editShortcut)
            )

            recyclerView?.adapter?.notifyItemInserted(presetListAdapters.size)

            saveVirtualControllerPresets()
        }

        fun deleteVirtualControllerPreset(name: String) {
            val index = presetList.indexOfFirst { it.name == name }

            presetList.removeAt(index)
            presetListAdapters.removeAt(index)

            recyclerView?.adapter?.notifyItemRemoved(index)

            if (index == selectedPresetId) {
                preferences?.edit {
                    putString(SELECTED_VIRTUAL_CONTROLLER_PRESET, presetListAdapters.firstOrNull()?.titleSettings)
                    apply()
                }
                recyclerView?.adapter?.notifyItemChanged(0)
            }

            saveVirtualControllerPresets()
        }

        fun renameVirtualControllerPreset(name: String, newName: String) {
            val index = presetList.indexOfFirst { it.name == name }

            presetList[index].name = newName
            presetListAdapters[index].titleSettings = newName

            recyclerView?.adapter?.notifyItemChanged(index)

            saveVirtualControllerPresets()
        }

        fun importVirtualControllerPreset(context: Context, file: File): Boolean {
            val lines = file.readLines()
            if (lines.size < 2) return false

            val type = lines[0]
            if (type != "virtualControllerPreset") return false

            val json = lines[1]
            val preset = gson.fromJson<VirtualControllerPreset>(json, listType)
            val canAutoAdjust = json.contains("resolution")

            var presetName = preset.name
            var count = 1

            while (presetList.any { it.name == presetName }) {
                presetName = "${preset.name}-${count++}"
            }

            preset.name = presetName

            if (canAutoAdjust) {
                val nativeResolution = getNativeResolution(context)

                if (preset.resolution != nativeResolution) {
                    val nativeSplit = nativeResolution.split("x").map { it.toFloat() }
                    val processedSplit = preset.resolution.split("x").map { it.toFloat() }

                    val multiplierX = nativeSplit[0] / processedSplit[0] * 100F
                    val multiplierY = nativeSplit[1] / processedSplit[1] * 100F

                    preset.buttons.forEach {
                        it.x = (it.x / 100F * multiplierX / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                        it.y = (it.y / 100F * multiplierY / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                    }
                    preset.analogs.forEach {
                        it.x = (it.x / 100F * multiplierX / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                        it.y = (it.y / 100F * multiplierY / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                    }
                    preset.dpads.forEach {
                        it.x = (it.x / 100F * multiplierX / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                        it.y = (it.y / 100F * multiplierY / GRID_SIZE).roundToInt() * GRID_SIZE.toFloat()
                    }
                }
            }

            presetList.add(preset)
            presetListAdapters.add(
                AdapterPreset.Item(preset.name, VIRTUAL_CONTROLLER_PRESET, true)
            )

            recyclerView?.post {
                recyclerView?.adapter?.notifyItemInserted(presetListAdapters.size)
            }

            saveVirtualControllerPresets()

            return true
        }

        fun exportVirtualControllerPreset(name: String, file: File) {
            val index = presetList.indexOfFirst { it.name == name }

            file.writeText("virtualControllerPreset\n" + gson.toJson(presetList[index]))
        }

        private fun saveVirtualControllerPresets() {
            preferences?.edit {
                putString("virtualControllerPresetList", gson.toJson(presetList))
                apply()
            }
        }

        fun getVirtualControllerPresets(): MutableList<VirtualControllerPreset> {
            val json = preferences?.getString("virtualControllerPresetList", "")
            val listType = object : TypeToken<MutableList<VirtualControllerPreset>>() {}.type

            return gson.fromJson(json, listType) ?: mutableListOf()
        }

        data class VirtualControllerPreset(
            var name: String,
            var resolution: String,
            var analogs: MutableList<VirtualKeyboardInputView.VirtualAnalog>,
            var buttons: MutableList<VirtualKeyboardInputView.VirtualButton>,
            var dpads: MutableList<VirtualKeyboardInputView.VirtualDPad>
        )
    }
}
